### DISCUSSION TITLE
nb2内存使用问题

### DISCUSSION THREAD

**Discussion by balaboooot** [nonebot/discussions#311] [问答] [answered]
半个萌新刚入坑，遇到了点问题，请大佬解答一下感激不尽 ！
服务器配置：E31265LV3 内存8G
1.请问nb2运行脚本是使用python的调试模式还是非调试模式
2.nb2下运行完插件后内存不会释放掉，导致python.exe内存占用越来越多
3.我自己检查了一下插件，发现并无内存泄漏之类的问题，因处理数据量大，单个插件非调试运行时会占用100-200MB的内存，运行完会立刻释放内存。
4.综合2.3点，bot每执行完一次指令，内存占用会从100MB到200MB，再到300MB……会直至爆满死机。爆满之前使用内存清理类的工具后（例如火绒、360等，不结束nb和python的进程）python.exe占用内存恢复正常。

插件是我自己编写的，要用numpy库处理数据
![}5~NAOZYE75~VDSHA}GJR~8](https://user-images.githubusercontent.com/102663004/216764988-35c1c043-cc53-465e-a83e-304296670a8a.png)

**Comment by yanyongyu** [comment, marked as answer]
nb自然不会用调试模式去运行脚本，nb本身只有存储插件创建的事件响应器，这些是整个bot运行的时候都存在的，也不会增加也不会减少。会出现增加的情况只有插件响应消息后进入会话状态且一直没有等到回复，通常会在会话超时（默认两分钟）后再次通过事件激活nb检查时释放。如果不是在state中存储了大量数据不会导致内存增长，你可以检查一下你的插件以及使用的第三方库是否进行了缓存，如jieba wordcloud等。基本可以排除是nb的问题（别人机器人运行几个月都没这问题）

**Reply by balaboooot** [reply]
请问您中间说的会话状态，是否与nb事件响应器中的.send()与.finish()指令有关，例如xx.finish()后，当前事件结束，脚本执行完成，释放内存，而如果是.send()的话，事件还在等待，默认两分钟后检查后才释放，不知我理解的是否正确
脚本我再检查一下，看是否存在问题

**Reply by yanyongyu** [reply]
是指`got` `receive`这些需要等待用户回复的操作，不是指向用户发送消息

**Reply by balaboooot** [reply]
懂了 目前我使用的都是receive，那我再检查一下，非常感谢您的解答

**Reply by yanyongyu** [reply]
即便你使用的receive也不会造成如此大的增长，除非你在state中存储了大量东西

**Reply by yanyongyu** [reply]
建议你找个memory profiler查下内存，是谁占着没放

**Reply by balaboooot** [reply]
找到问题了 应该是matplotlib库导致的内存泄露，非常感谢

**Reply by balaboooot** [reply]
顺便说一下这个问题吧，mpltb这个库，里面有个plt.show(),plt.savefig()等命令，在py自带的非调试模式下运行完毕后，内置的plt绘制的图也会跟着关闭，虽然没显示在窗口，导致你会误认为这个也一块关了，而nb中py虽然是非调试模式，但plt绘制的图一直堆积在内存中，所以show()和save()等命令后一定一定加上plt.close('all')这个命令行！

**Reply by balaboooot** [reply]
希望后面能帮助到遇到我这个问题的人
